### ISSUE TITLE
Deprecate SelfInitializedFakes 

### ISSUE THREAD

**Issue by blairconrad** [FakeItEasy/FakeItEasy#820]
... because we generally aren't excited about the feature.

And we're planning on removing it. See #798.

**Comment by adamralph** [comment]
See #798, there's some doubt about whether we want to do this.

**Comment by adamralph** [comment]
Based on recent discussions, I'm unblocking this and putting it on the 3.0.0 milestone.

**Comment by blairconrad** [comment]
I notice that we expose `IFakeOptionsForWrappers` and `IFakeOptionsForWrappers<T>` in the netstd build, and they only expose `RecordedBy`.
Shall we replace them with `IFakeOptions` (`<T>`) for the netstd build? If we'd remove it once we remove Self-Initializing Fakes, I think it makes sense to make the change now.

**Comment by blairconrad** [comment]
Also, want to pre-tweak the `Obsolete` message?

```c#
[Obsolete("Self-initializing Fake functionality will be moved out of FakeItEasy in the next major release.")]
```

to appear on all the `RecordedBy`s. I didn't think that forcing an error was warranted. Let the sticklers' build settings do that for them…

**Comment by thomaslevesque** [comment]
> Shall we replace them with `IFakeOptions` (`<T>`) for the netstd build? If we'd remove it once we remove Self-Initializing Fakes, I think it makes sense to make the change now.

👍 

> Self-initializing Fake functionality will be moved out of FakeItEasy in the next major release

Maybe something less definite than "the next major release"? Unless we're certain we want to remove it in 4.0.0.

> I didn't think that forcing an error was warranted.

Agreed. Forcing an error would have the same effect as removing the feature right now. As long as it's not actually removed, it should still be usable.

**Comment by adamralph** [comment]
👍 to replacing with `IFakeOptions` (`<T>`)

In NServiceBus and related packages, we do indeed [specify the version number when the item will be removed](https://docs.particular.net/nservicebus/upgrades/release-policy#deprecation). I think it's safe to say that we'll do it in 4.0. My suggestion:

> Self-initializing fakes will be removed in version 4.0.0.

For the reason @thomaslevesque states, I agree that the obsolete should be a warning. We could be even kinder and warn in 4.0.0, error in 5.0.0 and remove in 6.0.0 (as we do with NServiceBus) but I think it's overkill for FakeItEasy. At least, for such a little-used feature.

**Comment by blairconrad** [comment]
> Self-initializing fakes will be removed in version 4.0.0.

This was my initial impulse as well. I don't know why I phrased as "next major release". So... vague.

> We could be even kinder and warn in 4.0.0, error in 5.0.0 and remove in 6.0.0 

Not warn in 3.0.0, error in 4.0.0, and remove in 5.0.0?

Either way, I agree that it's overkill.

**Comment by adamralph** [comment]
Oh, sorry, yes, I had an off by one error 😄

**Comment by blairconrad** [comment]
This issue has been released in FakeItEasy 3.0.0:
https://github.com/FakeItEasy/FakeItEasy/releases/3.0.0
